File: app/src/main/java/org/androidforfun/framework/impl/MultiTouchHandler.java
/*
 *  Copyright (C) 2016 Mario Zechner
 *  This file is part of Framework for book Beginning Android Games.
 *
 *  This library is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License.
 */
package org.androidforfun.framework.impl;

import java.util.ArrayList;
import java.util.List;

import android.view.MotionEvent;
import android.view.View;

import org.androidforfun.framework.Input.TouchEvent;
import org.androidforfun.framework.Pool;
import org.androidforfun.framework.Pool.PoolObjectFactory;

/*
 * This is the handler used to manage touch events on newer Android versions.
 *
 * @author mzechner
 */
public class MultiTouchHandler implements TouchHandler {
    boolean[] isTouched = new boolean[20];
    int[] touchX = new int[20];
    int[] touchY = new int[20];
    Pool<TouchEvent> touchEventPool;
    List<TouchEvent> touchEvents = new ArrayList<>();
    List<TouchEvent> touchEventsBuffer = new ArrayList<>();
    float scaleX;
    float scaleY;

    public MultiTouchHandler(View view, float scaleX, float scaleY) {
        PoolObjectFactory<TouchEvent> factory = new PoolObjectFactory<TouchEvent>() {
            @Override
            public TouchEvent createObject() {
                return new TouchEvent();
            }
        };
        touchEventPool = new Pool<>(factory, 100);
        view.setOnTouchListener(this);

        this.scaleX = scaleX;
        this.scaleY = scaleY;
    }

    @Override
    public boolean onTouch(View v, MotionEvent event) {
        synchronized (this) {
            int action = event.getAction() & MotionEvent.ACTION_MASK;
            int pointerIndex = (event.getAction() & MotionEvent.ACTION_POINTER_ID_MASK) >> MotionEvent.ACTION_POINTER_ID_SHIFT;
            int pointerId = event.getPointerId(pointerIndex);
            TouchEvent touchEvent;

            switch (action) {
            case MotionEvent.ACTION_DOWN:
            case MotionEvent.ACTION_POINTER_DOWN:
                touchEvent = touchEventPool.newObject();
                touchEvent.type = TouchEvent.TOUCH_DOWN;
                touchEvent.pointer = pointerId;
                touchEvent.x = touchX[pointerId] = (int) (event
                        .getX(pointerIndex) * scaleX);
                touchEvent.y = touchY[pointerId] = (int) (event
                        .getY(pointerIndex) * scaleY);
                isTouched[pointerId] = true;
                touchEventsBuffer.add(touchEvent);
                break;

            case MotionEvent.ACTION_UP:
            case MotionEvent.ACTION_POINTER_UP:
            case MotionEvent.ACTION_CANCEL:
                touchEvent = touchEventPool.newObject();
                touchEvent.type = TouchEvent.TOUCH_UP;
                touchEvent.pointer = pointerId;
                touchEvent.x = touchX[pointerId] = (int) (event
                        .getX(pointerIndex) * scaleX);
                touchEvent.y = touchY[pointerId] = (int) (event
                        .getY(pointerIndex) * scaleY);
                isTouched[pointerId] = false;
                touchEventsBuffer.add(touchEvent);
                break;

            case MotionEvent.ACTION_MOVE:
                int pointerCount = event.getPointerCount();
                for (int i = 0; i < pointerCount; i++) {
                    pointerIndex = i;
                    pointerId = event.getPointerId(pointerIndex);

                    touchEvent = touchEventPool.newObject();
                    touchEvent.type = TouchEvent.TOUCH_DRAGGED;
                    touchEvent.pointer = pointerId;
                    touchEvent.x = touchX[pointerId] = (int) (event
                            .getX(pointerIndex) * scaleX);
                    touchEvent.y = touchY[pointerId] = (int) (event
                            .getY(pointerIndex) * scaleY);
                    touchEventsBuffer.add(touchEvent);
                }
                break;
            }

            return true;
        }
    }

    @Override
    public boolean isTouchDown(int pointer) {
        synchronized (this) {
            return !(pointer < 0 || pointer >= 20) && isTouched[pointer];
        }
    }

    @Override
    public int getTouchX(int pointer) {
        synchronized (this) {
            if (pointer < 0 || pointer >= 20)
                return 0;
            else
                return touchX[pointer];
        }
    }

    @Override
    public int getTouchY(int pointer) {
        synchronized (this) {
            if (pointer < 0 || pointer >= 20)
                return 0;
            else
                return touchY[pointer];
        }
    }

    @Override
    public List<TouchEvent> getTouchEvents() {
        synchronized (this) {
            int len = touchEvents.size();
            for (int i = 0; i < len; i++)
                touchEventPool.free(touchEvents.get(i));
            touchEvents.clear();
            touchEvents.addAll(touchEventsBuffer);
            touchEventsBuffer.clear();
            return touchEvents;
        }
    }
}

File: app/src/main/kotlin/org/code4projects/framework/impl/MultiTouchHandler.kt
/*
 * Copyright (c) 2016-2026 Salvatore D'Angelo
 * Licensed under the MIT License. See the LICENSE file in the project root.
 */
package org.code4projects.framework.impl

import android.view.MotionEvent
import android.view.View

import org.code4projects.framework.Input.TouchEvent
import org.code4projects.framework.Pool
import org.code4projects.framework.Pool.PoolObjectFactory

/*
 * This is the handler used to manage touch events on newer Android versions.
 *
 * @author mzechner
 */
class MultiTouchHandler(view: View, private val scaleX: Float, private val scaleY: Float) : TouchHandler {
    private val isTouched = BooleanArray(20)
    private val touchX = IntArray(20)
    private val touchY = IntArray(20)
    private val touchEventPool: Pool<TouchEvent>
    private val touchEvents: MutableList<TouchEvent> = ArrayList()
    private val touchEventsBuffer: MutableList<TouchEvent> = ArrayList()

    init {
        val factory = object : PoolObjectFactory<TouchEvent> {
            override fun createObject(): TouchEvent = TouchEvent()
        }
        touchEventPool = Pool(factory, 100)
        view.setOnTouchListener(this)
    }

    override fun onTouch(v: View?, event: MotionEvent): Boolean {
        synchronized(this) {
            val action = event.action and MotionEvent.ACTION_MASK
            var pointerIndex =
                (event.action and MotionEvent.ACTION_POINTER_ID_MASK) shr MotionEvent.ACTION_POINTER_ID_SHIFT
            var pointerId = event.getPointerId(pointerIndex)
            var touchEvent: TouchEvent

            when (action) {
                MotionEvent.ACTION_DOWN, MotionEvent.ACTION_POINTER_DOWN -> {
                    touchEvent = touchEventPool.newObject()
                    touchEvent.type = TouchEvent.TOUCH_DOWN
                    touchEvent.pointer = pointerId
                    touchX[pointerId] = (event.getX(pointerIndex) * scaleX).toInt()
                    touchEvent.x = touchX[pointerId]
                    touchY[pointerId] = (event.getY(pointerIndex) * scaleY).toInt()
                    touchEvent.y = touchY[pointerId]
                    isTouched[pointerId] = true
                    touchEventsBuffer.add(touchEvent)
                }

                MotionEvent.ACTION_UP, MotionEvent.ACTION_POINTER_UP, MotionEvent.ACTION_CANCEL -> {
                    touchEvent = touchEventPool.newObject()
                    touchEvent.type = TouchEvent.TOUCH_UP
                    touchEvent.pointer = pointerId
                    touchX[pointerId] = (event.getX(pointerIndex) * scaleX).toInt()
                    touchEvent.x = touchX[pointerId]
                    touchY[pointerId] = (event.getY(pointerIndex) * scaleY).toInt()
                    touchEvent.y = touchY[pointerId]
                    isTouched[pointerId] = false
                    touchEventsBuffer.add(touchEvent)
                }

                MotionEvent.ACTION_MOVE -> {
                    val pointerCount = event.pointerCount
                    for (i in 0 until pointerCount) {
                        pointerIndex = i
                        pointerId = event.getPointerId(pointerIndex)

                        touchEvent = touchEventPool.newObject()
                        touchEvent.type = TouchEvent.TOUCH_DRAGGED
                        touchEvent.pointer = pointerId
                        touchX[pointerId] = (event.getX(pointerIndex) * scaleX).toInt()
                        touchEvent.x = touchX[pointerId]
                        touchY[pointerId] = (event.getY(pointerIndex) * scaleY).toInt()
                        touchEvent.y = touchY[pointerId]
                        touchEventsBuffer.add(touchEvent)
                    }
                }
            }

            return true
        }
    }

    override fun isTouchDown(pointer: Int): Boolean {
        synchronized(this) {
            return !(pointer < 0 || pointer >= 20) && isTouched[pointer]
        }
    }

    override fun getTouchX(pointer: Int): Int {
        synchronized(this) {
            return if (pointer < 0 || pointer >= 20) 0 else touchX[pointer]
        }
    }

    override fun getTouchY(pointer: Int): Int {
        synchronized(this) {
            return if (pointer < 0 || pointer >= 20) 0 else touchY[pointer]
        }
    }

    override fun getTouchEvents(): List<TouchEvent> {
        synchronized(this) {
            val len = touchEvents.size
            for (i in 0 until len)
                touchEventPool.free(touchEvents[i])
            touchEvents.clear()
            touchEvents.addAll(touchEventsBuffer)
            touchEventsBuffer.clear()
            return touchEvents
        }
    }
}
